### PR TITLE
Generate command

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

Converts`CommandRegistry` (now called `TuistCommand`) and `GenerateCommand` to new argument parser. More info [here](https://github.com/tuist/tuist/pull/1154)

### Implementation 👩‍💻👨‍💻

- [x] Rewrite `CommandRegistry`
- [x] Rewrite `GenerateCommand`
- [x] Rewrite `GenerateCommandTests` 

### Unsolved

Since we are rewriting `CommandRegistry`, but have not implemented yet all commands, feature tests will fail for a while now. But it will be checked before merging to master, so I deem it to be an acceptable limitation.

Also we have not implemented `RawCommands` and `HiddenCommands` as I have written about in the previous PR:
> Instead of CommandRegistry, I have introduced a TuistCommand and its subcommands are all commands from TuistKit. As far as I can tell, I have been able to keep the former behavior except rawcommand and hiddenCommand - but hidden commands are not being used currently and RawCommand is only for BuildCommand, so I wonder if we need to come up with a solution for these (probably @pepibumur can answer this). Otherwise I have had to change ErrorHandler as main() returns Never - to best preserve our current setup, I am now exiting right from TuistCommand

### CODE REVIEW

File: Sources/TuistKit/Commands/GenerateCommand.swift [hunk 1]
```diff
@@ -1,70 +1,28 @@
...
+    
+    @Option(
+        name: .shortAndLong,
+        default: false,
+        help: "Only generate the local project (without generating its dependencies)."
+    )
+    var projectOnly: Bool
+    
+    func run() throws {
+        try GenerateService().run(path: path,
```
Comment (pepibumur) [inline Sources/TuistKit/Commands/GenerateCommand.swift]: Beautiful!

File: Sources/TuistKit/Commands/GenerateCommand.swift [hunk 2]
```diff
@@ -1,70 +1,28 @@
...
+        help: "The path where the project will be generated."
+    )
+    var path: String?
+    
+    @Option(
+        name: .shortAndLong,
+        default: false,
+        help: "Only generate the local project (without generating its dependencies)."
+    )
+    var projectOnly: Bool
```
Comment (pepibumur) [inline Sources/TuistKit/Commands/GenerateCommand.swift]: Where do you define the name of the option? Is it inferred automatically?

File: Sources/TuistKit/Commands/TuistCommand.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import Foundation
+import ArgumentParser
+import TuistSupport
+
+public struct TuistCommand: ParsableCommand {
+    public init() { }
+    
+    public static var configuration: CommandConfiguration {
```
Comment (pepibumur) [inline Sources/TuistKit/Commands/TuistCommand.swift]: Is this required by the `ParsableCommand` protocol?

File: Sources/TuistKit/Commands/GenerateCommand.swift (hunk 2 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/GenerateCommand.swift]: Yes, it's kind of magic 🎩 (including the `shortVersion`)

File: Sources/TuistKit/Commands/TuistCommand.swift (hunk 3 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/TuistCommand.swift]: Yes.

File: Sources/TuistKit/Commands/TuistCommand.swift [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    
+    public static func main(_ arguments: [String]? = nil) -> Never {
+        let errorHandler = ErrorHandler()
+        let command: ParsableCommand
+        do {
+             command = try parseAsRoot(processArguments(arguments))
+        } catch {
+            logger.error("\(fullMessage(for: error))")
+            _exit(exitCode(for: error).rawValue)
```
Comment (pepibumur) [inline Sources/TuistKit/Commands/TuistCommand.swift]: Where is `_exit` defined?

File: Tests/TuistKitTests/Services/GenerateServiceTests.swift [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+ import TuistLoader
+ import TuistSupport
+ import XcodeProj
+ import XCTest
+ @testable import TuistCoreTesting
+ @testable import TuistKit
+ @testable import TuistLoaderTesting
+ @testable import TuistSupportTesting
+
+ final class GenerateServiceTests: TuistUnitTestCase {
```
Comment (pepibumur) [inline Tests/TuistKitTests/Services/GenerateServiceTests.swift]: I really like that we are extracting the logic from the commands into services. Nicely done @fortmarek

File: Sources/tuist/main.swift [hunk 6]
```diff
@@ -9,5 +9,4 @@ if CommandLine.arguments.contains("--verbose") {
 LogOutput.bootstrap()
 
 import TuistKit
-var registry = CommandRegistry()
-registry.run()
+TuistCommand.main()
```
Comment (pepibumur) [inline Sources/tuist/main.swift]: Does this mean that only the generate command works?

File: Sources/TuistSupport/Errors/ErrorHandler.swift [hunk 7]
```diff
@@ -12,27 +12,8 @@ public protocol ErrorHandling: AnyObject {
...
-    public convenience init() {
-        self.init(exiter: { exit($0) })
-    }
-
-    /// Default error handler initializer.
-    ///
-    /// - Parameters:
-    ///   - exiter: Closure to exit the execution.
-    init(exiter: @escaping (Int32) -> Void) {
-        self.exiter = exiter
```
Comment (pepibumur) [inline Sources/TuistSupport/Errors/ErrorHandler.swift]: If exiter is no longer necessary I think we can get rid of it.

File: Sources/TuistKit/Services/GenerateService.swift [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import TuistGenerator
+import TuistLoader
+import TuistSupport
+import Basic
+
+class GenerateService {
```
Comment (pepibumur) [inline Sources/TuistKit/Services/GenerateService.swift]: ❤️

File: Sources/TuistKit/Commands/GenerateCommand.swift (hunk 2 above)
Comment (pepibumur) [reply Sources/TuistKit/Commands/GenerateCommand.swift]: magic

File: Sources/tuist/main.swift (hunk 6 above)
Comment (fortmarek) [reply Sources/tuist/main.swift]: Yes - it'd be too messy to keep merging the migration into `master`, that's why this PR points to `argument-parser` branch - I have tried to explain it more clearly [here](https://github.com/tuist/tuist/pull/1154#issue-394363041) in `Implementation Strategy`

Comment (fortmarek) [reply Sources/tuist/main.swift]: that also means `feature` tests will not pass, but we will make sure those work before merging into master

File: Sources/TuistKit/Commands/TuistCommand.swift (hunk 4 above)
Comment (fortmarek) [reply Sources/TuistKit/Commands/TuistCommand.swift]: It's defined in `Darwin`- I have tried to copy `parseOrExit(_ arguments: [String]? = nil)` defined in the parser library that we'd use if we did not need custom error handling

File: Sources/TuistSupport/Errors/ErrorHandler.swift (hunk 7 above)
Comment (fortmarek) [reply Sources/TuistSupport/Errors/ErrorHandler.swift]: That's what this change does 🙂

**Comment by fortmarek** [comment]
Thanks for the comments @pepibumur! I really like working with the argument parser, so far the migration has been a joy 🙂 Once you give this a stamp of approval, I'll merge into `argument-parser` branch.

**Pull request merged**